pyrefly/lib/commands/config_migration.rs

```diff
@@ -164,12 +164,69 @@ impl Args {
-    let config = PyProject::new(config);
-    let serialized = toml::to_string_pretty(&config)?;
-    fs_anyhow::append(pyproject_path, serialized.as_bytes())
-        .with_context(|| "While trying to write the pyrefly config to the pyproject.toml file")
+    const ERR_WRITE_CONFIG: &str = "While trying to write pyrefly config to pyproject.toml file";
+
+    if pyproject_path.exists() {
+        let original_content = fs_anyhow::read_to_string(pyproject_path)?;
+        let doc = original_content
+            .parse::<toml_edit::DocumentMut>()
+            .with_context(|| {
+                format!(
+                    "Failed to parse {} as TOML document",
+                    pyproject_path.display()
+                )
+            })?;
+
+        let config_pyproject = PyProject::new(config.clone());
+        let toml = toml::to_string_pretty(&config_pyproject)?;
+        let pyrefly_section = if let Some(start) = toml.find("[tool.pyrefly]") {
+            toml[start..].trim_end().to_owned()
+        } else {
+            String::from("[tool.pyrefly]")
+        };
+        let mut updated_content = doc.to_string();
+        updated_content = updated_content.trim_end().to_owned();
+        while let Some(start) = updated_content.find("[tool.pyrefly]") {
+            let after = &updated_content[start + 1..];
+            let next_section = after.find('\n').and_then(|nl| {
+                let rest = &after[nl + 1..];
+                rest.find("\n[").map(|i| nl + 1 + i)
+            });
+            let end = next_section
+                .map(|i| start + 1 + i + 1)
+                .unwrap_or(updated_content.len());
+            let mut cleaned = String::new();
+            cleaned.push_str(&updated_content[..start]);
+            cleaned.push_str(&updated_content[end..]);
+            updated_content = cleaned.trim_end().to_owned();
+        }
+        if updated_content.is_empty() {
+            updated_content = pyrefly_section;
+        } else {
+            updated_content.push_str("\n\n");
+            updated_content.push_str(&pyrefly_section);
+        }
+        updated_content.push('\n');
+        fs_anyhow::write(pyproject_path, updated_content.as_bytes())
+            .with_context(|| ERR_WRITE_CONFIG)
+    } else {
+        let config_pyproject = PyProject::new(config);
+        let mut serialized = toml::to_string_pretty(&config_pyproject)?;
+        if !serialized.contains("[tool.pyrefly]") {
+            serialized = String::from("[tool.pyrefly]\n");
+        } else {
+            serialized = serialized.trim().to_owned();
+            serialized = format!("{serialized}\n");
+            if serialized
+                .trim_start_matches('\n')
+                .starts_with("[tool.pyrefly]")
+                && !serialized.trim_start_matches('\n').contains('\n')
+            {
+                serialized = String::from("[tool.pyrefly]\n");
+            }
+        }
```

Comment:
Similar to above, can we create a new TOML document, add Pyrefly to it, and serialize it to disk?

Reply:
Updated in 0592ddfab153b5fc1ed49dfc63d7816a3741b78e.